### ISSUE TITLE
Igniter installers

### ISSUE THREAD

**Issue by tmjoen** [brandocms/brando#2462]
Make Igniter the canonical foundation for Brando installation, code generators, and source upgrades, with a reliable path from a fresh Phoenix project to the first working content type. Target the 0.55 work discussed in the existing comment.

## Closure — 2026-09-13

The original installer and maintained source-generator implementation is complete and merged into `next` through PR #2785. All remaining delivery gates are now owned by [#2823 — Prepare 0.55.0-beta.0: coordinated Hex and npm release](https://github.com/brandocms/brando/issues/2823): supported consumer compatibility, actual package-version upgrades, and coordinated package publication, artifact qualification and published documentation.

Closing this implementation issue does not claim those release gates have passed. The unchecked release criteria below remain as historical scope and are tracked to completion in #2823. Broader consumer versions/layouts must be qualified there before being advertised.

## Implementation progress — 2026-09-07

The four original tasks and the maintained source-generator migration are **merged into `next`** through **[PR #2785](https://github.com/brandocms/brando/pull/2785)**, merge commit [`f69d779b5`](https://github.com/brandocms/brando/commit/f69d779b536806221b8ed78671c1f20e2069ebb4), on 2026-09-07. The reviewed head was `164115384`. Checked boxes describe completed implementation and qualification; they do not claim the published package contains this work.

- [x] One native `brando.install` implementation, exercised directly and through real `igniter.install brando@path:...` composition. Shared source discovery, validated options, semantic edits and owned-file conflicts preserve application customization.
- [x] New installations default to tenancy `none`; reruns preserve existing settings. Common opt-in `--interactive` guides missing choices and respects explicit arguments. `--yes` accepts the diff without authorizing ownership changes or overwrites.
- [x] Cold one-shot `igniter.new --with phx.new --install brando@path:...` installation and compilation pass from outside an existing Mix project. Previously compiled consumers also install successfully after adding Igniter, without a forced rebuild. Guarded tasks and helpers use Mix's optional-dependency recompilation hook; subprocess tests cover both dependency transitions.
- [x] Native frontend/backend generators preserve dependency sources. Explicit Yalc setup builds both consumer Vite projects; binary assets use digest-checked, exact-byte deferred copying. Keep Yalc during iteration.
- [x] Argument-driven Blueprint generation uses the current DSL, naming overrides and consumer template precedence. Resource generation reads compiled metadata, extends contexts semantically, integrates admin routes and optionally creates public rendering routes. Multiple resources compile in a shared context without replacing custom functions.
- [x] Optional CMS public-site scaffolding is available through `--public-site` or `brando.gen.site`. It uses separate CMS modules, tenant/locale/identity setup and published-page rendering. Replacing the stock Phoenix homepage requires `--replace-phoenix-home` or an affirmative interactive answer; its source files remain. Other homepage/catch-all ownership blocks for explicit integration.
- [x] Blueprint migration/snapshot planning preserves locking, stale-plan/history checks, deterministic cross-process verification, no-clobber writes and rollback on snapshot failure. Real previews/declines leave storage unchanged; accepted pairs apply successfully. Framework and tenant migration-file generators use the same reviewed source workflow.
- [x] Maintained auxiliary generators now use Igniter: mail, sitemap, authorization, release and tenant migrations. Telemetry defaults, discovered application/Repo identity, tenancy guidance and the 0.54 migration are hardened. Fabric scaffolding and `install.sh` have documented retirement paths. Database operations, builds and deployment remain explicit execution steps.
- [x] Native upgrade compatibility bridge: a recognized legacy consumer task is archived through a reviewed plan, a separate compile removes its stale beam, and Igniter resolves the library's version-checked hook. Customized tasks block removal. Real package-version transition qualification remains a release gate.
- [x] Consumer qualification exposed and fixed shared-public-user foreign keys, tenant migration destinations/history checks, tenant-aware public controllers and shared-user cascades before the first environment. Source guards reject conflicting admin/authentication ownership, embedded Blueprint resources and overlapping generated context names.
- [x] **169 focused source/storage tests and 7 tenant-job database tests passed.** Warnings-as-errors compilation, formatting and shell syntax checks passed. Configured code analysis found no findings on changed lines; existing repository findings remain.
- [x] **Four complete consumer scenarios passed locally:** fresh `none`, `single`, `multi`, and a customized/precompiled `none` application. They cover preview/rerun fingerprints, exact binary assets, both Vite builds, migrations, provisioning, login and generated-resource create/edit/public rendering. CMS admin edits update the public homepage in all three modes; drafts, 404s and support routes are checked. The customized app retains files, routes, configuration, context functions and supervision.
- [x] Reproducible smoke script and four-case GitHub Actions matrix added. All four cases [passed remotely on the initial revision](https://github.com/brandocms/brando/actions/runs/34072443622). All four cases also [passed after the CI fixes](https://github.com/brandocms/brando/actions/runs/34073198244). Repository checks passed all 2,158 tests on Elixir 1.17.2, 1.18.1, 1.19.5 and 1.20.3 with their configured OTP versions, plus documentation, Blueprint quality and authorization E2E. The pre-review revision `ce23a536f` also passed the [full unit matrix, documentation and Blueprint quality checks](https://github.com/brandocms/brando/actions/runs/34073929763) and [all four consumer cases](https://github.com/brandocms/brando/actions/runs/34073929760). All three broader E2E jobs also completed successfully on that pre-review revision: legacy-1 had 89 passed, legacy-2 had 77 passed and 11 skipped, and authorization plus its compatibility checks had 43 passed. The test steps and Playwright summaries were verified directly; this result does not rely on the jobs being nonblocking.

- [x] Fixed issues exposed by repository CI: explicit documentation for compatibility delegates; Elixir 1.19 snapshot type matching; reloading cached base-project source in Igniter fixtures; current tenant-migration argument tests; isolating the legacy upgrade template test; and preserving the E2E consumer's legacy task outside compilation paths. Migration execution tests now seed and verify the correct shared public user reference. Local source, migration, documentation and Blueprint quality checks pass after these changes.

**Verified local combination:** Phoenix/phx_new 1.8.13, LiveView 1.2.11, Elixir 1.20.3, OTP 28.4.1, Igniter 0.8.3, igniter_new 0.5.28, PostgreSQL 16.1, Node.js 23.11.0, pnpm 10.4.1 and Yalc 1.0.0-pre.53. The remote consumer matrix also passed on PostgreSQL 14, Node.js 22 and pnpm 10.32.1. Standalone Phoenix applications are the supported initial layout; broader layouts/version combinations need dedicated coverage before being advertised.

- [x] Focused review completed for installer ownership, migration/snapshot consistency, legacy upgrade compatibility and guided setup. Fixed parameterized route overlap detection and authentication-table detection for `create_if_not_exists` in `164115384`. Regression tests reproduced both gaps; all 26 affected tests pass after the fixes.
- [x] Final review revision `164115384` passed [repository PR CI](https://github.com/brandocms/brando/actions/runs/34093830250), [all four consumer cases](https://github.com/brandocms/brando/actions/runs/34093830220), and the separate [push CI](https://github.com/brandocms/brando/actions/runs/34093826160). The unit matrix passed 2,159 tests, including 135 doctests. Actual E2E steps and logs show 209 passed, 11 skipped, zero failures (89 legacy-1, 77 legacy-2, 43 authorization/compatibility). Documentation and Blueprint quality checks passed. PR #2785 is merged into `next`.

### Remaining delivery plan — transferred to #2823

The following gates are now tracked in #2823 and do not remain open work in this implementation issue.

1. **Supported consumer versions:** confirm the release-supported consumer version matrix and qualify any broader layouts/version combinations before advertising them. The initial documented standalone Phoenix combination is qualified and merged.
2. **Real version upgrades:** qualify an actual old-consumer/package-version transition through `mix igniter.upgrade brando`, including historical migrations, customization and exact database follow-ups. The bridge and dispatcher are tested; an equal-version `0.54.0-dev` hook is not evidence for a future `0.55` recipe.
3. **Release artifacts and npm, near release:** publish the matching BrandoJS package, establish its Brando version pairing and make npm the released installer's default. Repeat clean install, existing-app, resource and upgrade checks against packaged artifacts without a sibling checkout or Yalc. Verify published documentation against those exact artifacts.

The original onboarding implementation is merged into `next`. #2823 owns these outstanding release gates; npm publication remains deliberately deferred until release preparation.

## Audited status

Audited against `next` at [`c31f791e4`](https://github.com/brandocms/brando/commit/c31f791e403c9456a4ea7d28ed4aedb1c6770802) on 2026-09-06. At that audit none of the original four checkboxes was complete. The implementation progress above records subsequent branch work; the three pre-existing Igniter tasks alone did not implement the headline workflow.

### Already implemented at the original audit

- [x] Add Igniter as an optional, non-runtime dependency (`~> 0.8.0`) and provide actionable missing-dependency instructions. See [`mix.exs`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/mix.exs#L244) and [`Mix.Brando.missing_igniter!/1`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/brando.ex#L371).
- [x] Implement `brando.gen.otel` using Igniter. This checks adoption only: project-specific defaults and repeat-run behavior still need work. See [`brando.gen.otel.ex`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/tasks/brando.gen.otel.ex).
- [x] Implement `brando.migrate54` using Igniter, with focused rewrite tests. See [task](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/tasks/brando.migrate.54.ex) and [tests](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/test/mix/tasks/brando_migrate54_test.exs).
- [x] Implement `brando.setup.tenancy` using Igniter, with argument validation and repeat-run coverage. See [task](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/tasks/brando.setup.tenancy.ex) and [tests](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/test/mix/tasks/brando_setup_tenancy_test.exs).

### Original scope

- [x] `igniter.install brando`: install and configure Brando through Igniter, including dependency resolution and composition.
- [x] `brando.install`: replace the current destructive scaffold task with the same Igniter installer used by `igniter.install brando`.
- [x] `brando.gen.blueprint`: replace the prompt-only template copier with an Igniter-native generator.
- [x] `brando.gen`: replace the prompt-only resource generator and marker-based context insertion with Igniter-native generation and AST edits.

`igniter.install brando` and `brando.install` are two entry points to one installer, not separate implementations. Igniter discovers a package's `<package>.install` task; see the [upstream installer conventions](https://github.com/ash-project/igniter#installers).

## Findings that motivated the migration (original audit)

- [`brando.install`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/tasks/brando.install.ex#L275) deletes `assets/`, deletes Phoenix templates, attempts `File.rename("test", "test/unit")`, and force-writes application/configuration files. At the audit, these operations bypassed any Igniter change plan; the native implementation above replaces this behavior.
- The generated [`mix.exs`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/priv/templates/brando.install/mix.exs#L80) hardcodes `{:brando, path: "../../brando"}`; the backend [package manifest](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/priv/templates/brando.install/assets/backend/package.json) uses Yalc. Yalc remains the intended workflow during heavy iteration. Preserve selected dependency sources and avoid hardcoding one directory layout; switch the released JS dependency to npm closer to release. npm publication is not a prerequisite for implementing or testing the Igniter tasks.
- [`brando.gen`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/tasks/brando.gen.ex) reads compiled Blueprint metadata, then force-copies templates and patches context text through [`Mix.Brando.add_to_file/4`](https://github.com/brandocms/brando/blob/c31f791e403c9456a4ea7d28ed4aedb1c6770802/lib/mix/brando.ex#L273). Routes are printed for manual insertion.
- The audited telemetry generator contained `by`/`BielkeYang` defaults and a hardcoded `Brando.O11y.OpenTelemetryEcto` setup call despite generating an application-namespaced module. Existing Igniter usage does not establish correctness or idempotence.

## Implementation sequence

This is the delivery map retained from the accepted plan. Steps 1–7 and 9 are implemented, qualified and merged into `next`. Step 8 has its compatibility bridge and hook; actual package-version transition qualification remains. Step 10 awaits release preparation. The source, consumer, documentation, quality and all three E2E checks passed on final review revision `164115384`. Focused review and merging are complete. The detailed requirements below distinguish those gates.

| Step | Depends on | Deliverable | Exit criterion |
| --- | --- | --- | --- |
| 1. Shared foundation and test harness | — | Introduce shared project discovery, validated options, template/conflict handling, and Igniter test fixtures. Record the supported versions/layouts and the public CLI contract. Cover the optional-dependency bootstrap. | Fixtures cover a fresh Phoenix app, a customized app, different app/module names, ambiguous Repos/routers, and closed stdin. Unsupported selections fail before scaffold writes. |
| 2. Core installer | 1 | Convert `brando.install`; compose focused operations for dependencies/configuration, supervision, endpoint/router/admin, owned templates, and required migration files. Use AST edits for existing Elixir code. | Direct and composed task invocations produce equivalent changes. Source installation works without a running database; applying and rerunning does not duplicate code or regenerate secrets. |
| 3. Assets and working fresh install | 2 | Keep the current Yalc workflow while iterating. Make dependency source selection explicit and preserve the chosen Brando/BrandoJS inputs. Connect the development bootstrap to consumer asset builds and explicit initialization steps; add the fresh-consumer smoke job now. | A fresh consumer using documented local Brando/Yalc inputs from the same framework revision builds both asset consumers, migrates a disposable database, initializes the selected tenancy mode, and reaches admin login. This step does not depend on npm publication. |
| 4. Existing Phoenix app support | 3 | Add deliberate public-site/route selection and asset conflict handling. Preserve customized dependencies, application children, routes, templates, tests, and configuration. | Installation and a second run preserve a customized consumer. Collisions produce actionable instructions; preview/decline does not write Brando scaffold files. Complete the two original installer checkboxes here. |
| 5. Blueprint generator | 1 | Convert `brando.gen.blueprint` to structured arguments and current templates; reuse discovery and conflict handling. Keep optional interactive guidance and template override support. | `Catalog Product` generates a compilable Blueprint in a differently named consumer; invalid names, naming overrides, existing files, and unattended invocation are covered. Complete the original Blueprint checkbox here. |
| 6. Resource generator | 4, 5 | Convert `brando.gen` to compiled metadata → generation plan → reviewed changes. Replace context text markers with AST edits and integrate admin routes. | Generate two resources into one context; compile their context/web/admin/test files; preserve custom code on reruns. Extend consumer smoke coverage to create/edit/render a generated resource. Complete the original resource checkbox here. |
| 7. Migration and snapshot planning | 6 | Separate storage-diff planning from persistence; integrate reviewed migration/snapshot output and framework migration-file generation while retaining Brando's existing storage rules. | Tests cover no-op, cancellation, stale plans/concurrent changes, failure during persistence, migration ordering, and explicit rebaseline. The adapter preserves the existing consistency guarantees. |
| 8. Native upgrade integration | 7 | Introduce a library-owned migration-file command with a distinct name, then migrate the recognized consumer-owned `brando.upgrade` task through a reviewed compatibility change. Add version-aware `brando.upgrade FROM TO` for Igniter. | A supported old-consumer fixture upgrades through `igniter.upgrade`, preserves historical migration files/customization, compiles, and receives exact database follow-up instructions. Customized legacy tasks get an actionable conflict instead of silent replacement. |
| 9. Remaining source tasks | 2, 7, 8 as applicable | Convert maintained auxiliary generators in small batches and harden the three existing Igniter tasks. Route all maintained source-writing commands through the shared contracts. | Each task has argument, composition, conflict, and rerun coverage; generated output uses consumer namespaces. Every legacy entry point has a supported implementation or documented replacement. |
| 10. npm transition and release qualification | 1–9 | Closer to release, publish BrandoJS to npm, make the compatible npm package the released installer's default, and document its pairing with Brando. Run the complete supported consumer matrix, finish the release/contributor guides, and retire `install.sh` and superseded scaffolding after the replacement paths pass. | CI installs the intended release artifacts in a clean workspace without Yalc or a sibling Brando checkout. A newcomer can follow the published commands through first login and a working content type without dependency-path edits or undocumented repair steps. |

The initial foundation, fresh-install, existing-app and content-generation milestones have now passed locally. Keep future source operations small and composable internally; expose additional public tasks only when they represent a useful user operation.

## Design decisions to carry into implementation

- **Opt-in guided setup:** use a common `--interactive` option across the installer and generators. Normal commands use documented defaults or explicit arguments without hidden prompts. Guided mode asks for missing choices, explains the options, validates responses, and respects arguments already supplied. Closed stdin must produce an actionable error. Keep `--yes` for accepting a change plan; it does not enable guided setup. Preserve compatibility with the existing tenancy prompt flags.
- **One installer:** both installation entry points call the same implementation. Fresh-project creation is supplied by `igniter.new` and `phx.new`.
- **Classic single-site by default:** on a new installation, omitting `--tenancy-mode` selects `none` without a tenancy prompt, in both direct and composed installation. On reruns, omission preserves existing tenancy configuration. Selecting `single` or `multi` requires an explicit mode; `single` also requires a valid `--site-key`. Keep `--tenancy-mode none` available for explicit scripts and preserve compatibility with the existing `--no-tenancy-prompt` flag.
- **Yalc now, npm near release:** keep Yalc during heavy iteration and use it in development consumer checks. Implement dependency source selection so the later npm transition changes package configuration without requiring an installer rewrite. npm publication gates release qualification, not the earlier Igniter implementation steps. Reassess whether a local Yalc option remains useful after the transition; indefinite Yalc support is not a requirement here.
- **Explicit ownership:** patch shared application files; create Brando-owned files; preserve customized files and report conflicts. Never use bulk deletion or forced copying as the conversion strategy.
- **Two content-generation stages initially:** accept the Blueprint, compile it, then generate the surrounding resource and storage changes. A future combined command needs a separate design for unpublished Blueprint metadata.
- **Separate planning from execution:** source generators plan changes; database provisioning, content creation, asset builds, and deployment remain explicit operations. Do not wrap a disk-writing legacy task inside `igniter/1`.
- **Prove compatibility before advertising it:** start with standalone Phoenix applications. Keep path/Git contributor workflows explicit and qualify umbrella support separately.
- **Test as each task lands:** use in-memory Igniter fixtures for source changes and generated-consumer checks for behavior. Build JS/CSS through consumer Vite projects, following the repository's asset validation boundary.

### 1. Establish the installation contract

- [x] Document the tested Phoenix/Elixir/PostgreSQL combination and supported initial project shape, with broader release-matrix qualification kept separate. Discover the OTP app, application/web/admin namespaces, Repo, router, and endpoint; allow explicit selection when ambiguous. Reject unsupported layouts before scaffold writes. Start with standalone Phoenix projects; advertise umbrella support only after dedicated coverage.
- [x] Make the normal bootstrap install/enable Igniter without requiring users to diagnose Brando's compile-time optional-dependency guard. Test both first installation and a previously compiled Brando dependency. Keep Igniter out of production runtime requirements.
- [x] Preserve the user's Brando dependency source and constraints and the selected BrandoJS source. Support and document the current local/Git + Yalc iteration workflow, using matching framework and JavaScript revisions. Do not require npm publication to complete the initial installer and generator work.
- [ ] At release qualification, establish the Brando/BrandoJS version pairing and default released installs to the published npm package. Verify that this release path requires neither a sibling framework checkout nor Yalc.

### 2. Implement one composable installer

- [x] Convert `brando.install` to `Igniter.Mix.Task` with `info/2`, structured options, and `igniter/1`. Compose focused setup operations for dependencies/configuration, supervision, endpoint/router/admin, assets, and migration files. Keep the existing direct task name.
- [x] Patch existing Elixir files semantically and preserve unrelated dependencies, configuration, routes, tests, and assets. Create owned templates with explicit conflict handling. Repeating installation must not duplicate code/migrations, rotate secrets, or overwrite customization. Eliminate bulk deletion and forced scaffold replacement.
- [x] Support both a documented fresh-project command using `igniter.new --with phx.new --install brando` and installation into a supported existing Phoenix app. Existing-app installation must resolve route/asset ownership explicitly; the optional Brando public-site scaffold must not silently claim an existing homepage.
- [x] Default a new installation to tenancy mode `none` when `--tenancy-mode` is absent, without prompting. Preserve validation for explicit `single`/`multi` modes and require a valid `--site-key` only for `single`; reject a site key without that explicit mode. On reruns, an omitted flag must preserve existing tenancy configuration rather than reset it to `none`. Cover defaults, explicit choices, reruns, closed stdin, and direct/composed invocation in tests. Retain compatibility with `--no-tenancy-prompt`.
- [x] Implement `--interactive` for guided installation, with equivalent CLI options for every required choice. Ask only for missing choices and validate answers; for example, `--interactive --tenancy-mode single` asks for the missing site key. Normal commands and noninteractive previews must not trigger hidden Brando prompts. Detect conflicts with actionable messages; `--yes` must not implicitly authorize replacing customized files or enable interactive mode.
- [x] Allow source setup without a running database; dependency fetching and project compilation may still be required. Keep database creation/migration, tenant provisioning, language records, administrator creation, and sample seeds as explicit ordered setup operations. A preview must never create an administrator or mutate content.

### 3. Implement the content-type workflow

- [x] Add argument-driven `brando.gen.blueprint Catalog Product` with inferred application namespace, explicit naming overrides, validation, and a guided `--interactive` mode for missing choices. Generate a current, compilable Blueprint using the supported DSL and Gettext conventions; preserve template customization with defined precedence and collision behavior.
- [x] Add `brando.gen Studio.Catalog.Product` with AST-based context extension and admin route integration. Preserve custom functions, avoid duplicate aliases/types/routes, handle namespaces and missing/invalid Blueprint metadata, and generate usable admin/web/test files. Public routes and authorization policy require explicit choices where they cannot be inferred.
- [x] Make the compilation boundary explicit. The current resource generator requires a compiled Blueprint; use an accepted Blueprint → compile → resource/migration workflow initially. Do not write pending Blueprint files or run the old generator behind Igniter's preview. Any later combined command needs shared generation metadata or another tested way to consume unpublished Blueprint changes.

### 4. Complete source-task adoption without weakening storage guarantees

- [x] Convert the maintained sitemap, mail, authorization, frontend/backend, release, and tenant-migration generators to the same preview/composition/conflict contract. Give the legacy Fabfile installer an explicit supported replacement or deprecation path. Retire `install.sh` once the supported Igniter bootstrap covers its necessary setup steps.
- [x] Adapt Blueprint migration/snapshot generation and framework migration-file generation to reviewed planning without replacing Brando's storage diff engine. Preserve locking, history validation, concurrent-write protection, migration/snapshot consistency, and explicit `--rebaseline` semantics. A cancelled preview must leave both files and snapshot versions unchanged. Do not claim Igniter's multi-file writes provide the existing guarantees automatically.
- [x] Bring the three existing Igniter tasks up to the same project-discovery and repeat-run standard, including removing hardcoded telemetry app/service identity and proving no duplicate setup calls.
- [x] Resolve the legacy consumer-owned `Mix.Tasks.Brando.Upgrade` naming collision through a reviewed archive/compile compatibility path. Add the library-owned version-checked `brando.upgrade FROM TO` hook, exercise Igniter dispatcher resolution and document path-dependency upgrades. Customized legacy tasks require explicit integration. See [Igniter's upgrade contract](https://github.com/ash-project/igniter/blob/main/documentation/upgrades.md).
- [ ] Qualify actual supported package-version transitions through `mix igniter.upgrade brando`, including the release-specific source recipe, historical migrations, compile compatibility and exact database follow-up instructions. The current dev-version bridge test does not replace this gate.
- [x] Keep execution tasks such as `ecto.migrate`, `brando.migrate_to_tenant`, administrator creation, builds, and deployment as explicit operational commands. Full Igniter adoption covers source changes; it does not require making database and deployment operations pretend to be source patches.

### 5. Prove the adoption experience

- [x] Add Igniter fixture tests for first run, repeat run, customized existing files, collisions, invalid arguments, module/Repo/router selection, supported layouts, missing Igniter, and closed stdin. Exercise both direct and composed entry points; verify declined/dry-run plans have no Brando scaffold, database, or snapshot side effects. Document any upstream dependency-fetch/compile behavior separately.
- [x] Add fresh-consumer CI during implementation using the documented local/Yalc inputs: bootstrap a Phoenix consumer, install Brando, compile, build frontend/backend assets in that consumer, migrate a disposable PostgreSQL database, initialize content in the documented order, log into admin, and create/edit/render a generated resource. Include tenancy modes and a customized existing-app fixture. All four cases pass locally on the exact combination recorded above.
- [x] Run the four-case consumer matrix remotely and verify the full unit suite across the four configured Elixir/OTP combinations. These checks passed on final review revision `164115384`; see the verified run links above.
- [ ] Qualify any broader consumer compatibility range selected for the release. The documented initial consumer combination and the unit-test matrix are distinct guarantees.
- [ ] Near release, run those consumer checks against the packaged Brando artifact and published npm package in a workspace without local framework/Yalc inputs. Verify package contents and version compatibility, not only repository templates.
- [x] Keep Yalc development onboarding accurate, with prerequisites, exact commands, configuration and ownership choices, generated changes, upgrade compatibility and recovery steps. The documented local workflow reaches first login and a working content type.
- [ ] Add tested npm-based release instructions during artifact qualification. The released workflow must reach first login without hand-editing generated dependency paths or undocumented repair commands.

## Intended command experience

The commands below are implemented and qualified with the explicit matching checkout and Yalc. Run one-shot project creation from outside an existing Mix project. Bare `igniter.install brando` is the intended packaged-release experience; that artifact qualification remains open. Replace the example absolute path with the actual checkout:

```sh
# After installing the documented Phoenix and Igniter archives:
mix igniter.new studio --with phx.new \
  --with-args="--no-install --no-assets --no-dashboard --no-mailer" \
  --install brando@path:/absolute/path/to/brando

# Or, in a supported existing Phoenix project:
mix igniter.install brando@path:/absolute/path/to/brando

# Or opt into guided installation:
mix igniter.install brando@path:/absolute/path/to/brando --interactive

# On a new install, tenancy defaults to none without a prompt.
# Choose one of these instead when named environments/tenancy are wanted:
mix igniter.install brando@path:/absolute/path/to/brando --tenancy-mode single --site-key studio
mix igniter.install brando@path:/absolute/path/to/brando --tenancy-mode multi

# Optional CMS pages with explicit stock-homepage ownership:
mix brando.gen.site --interactive

# In the installed consumer:
mix brando.gen.blueprint Catalog Product
# Define the content model, review and accept it, then:
mix compile
mix brando.gen Studio.Catalog.Product
mix brando.gen.blueprint_migration Studio.Catalog.Product
```

The reviewed source changes are followed by the documented asset/database/account setup. A successful generator command alone is not the completion criterion; a functioning consumer and preservation of existing application code are.

## Definition of done

The implementation criteria are complete. The remaining release criteria are transferred to #2823; they are left unchecked here to preserve the distinction between merged implementation and qualified release artifacts.

- [x] All four original checkboxes are complete and backed by direct/composed invocation tests.
- [x] All maintained source-generating tasks meet the preview, conflict, unattended-execution, and repeat-run contract, or have an explicit documented retirement path.
- [ ] Released installation, existing-app installation, generated-resource behavior, and supported upgrades pass the consumer checks above.
- [ ] Documentation and CLI help describe the shipped behavior, prerequisites, and follow-up commands; examples are verified against the packaged artifacts.

The original tasks and maintained source-generator adoption are complete and merged into `next` through PR #2785, so this implementation issue is closed. #2823 now owns supported release upgrades, broader consumer compatibility selected for release, packaged-artifact qualification, npm publication and verification of published documentation. Those release criteria must pass in #2823 before the corresponding release support is claimed.

**Comment by tmjoen** [comment]
This can wait to 0.55
